gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+      TimeAggregation agg =
+          TimeAggregation.newBuilder()
+              .setAggregation(aggregation)
+              .setPeriod(timeAggregation.getPeriod())
+              .build();
+      timeAggregations.add(agg);
+    }
+    return timeAggregations;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForAggregateExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<FunctionExpression> aggregateList = originalRequest.getBaselineAggregateRequestList();
+    List<TimeAggregation> timeAggregationList = new ArrayList<>();
+    for (FunctionExpression function : aggregateList) {
+      TimeAggregation timeAggregation =
+          getAggregationFunction(
+              function, originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      timeAggregationList.add(timeAggregation);
+    }
+    return timeAggregationList;
+  }
+
+  private TimeAggregation getAggregationFunction(
+      FunctionExpression function, long startTimeMillis, long endTimeMillis) {
+    Period period = getPeriod(startTimeMillis, endTimeMillis);
+    TimeAggregation timeAggregation =
+        TimeAggregation.newBuilder()
+            .setPeriod(period)
+            .setAggregation(Expression.newBuilder().setFunction(function).build())
+            .build();
+    return timeAggregation;
+  }
+
+  private Period getPeriod(long startTimeInMillis, long endTimeMillis) {
+    Period period;
+    if ((endTimeMillis - startTimeInMillis) < DAY_IN_MILLIS) {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60);
+      period = Period.newBuilder().setUnit("MINUTES").setValue((int) timePeriod).build();
+    } else {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60 * 60 * 24);
+      period = Period.newBuilder().setUnit("HOURS").setValue((int) timePeriod).build();
+    }
+    return period;
+  }
+
+  private Period getTimeSeriesPeriod(List<BaselineTimeAggregation> baselineTimeSeriesList) {
+    return baselineTimeSeriesList.get(0).getPeriod();
+  }
+
+  private long getPeriodInSecs(Period period) {
+    ChronoUnit unit = ChronoUnit.valueOf(period.getUnit());
+    long periodSecs = Duration.of(period.getValue(), unit).getSeconds();
+    return periodSecs;
+  }
+
+  /**
+   * If user selected time range is less than 24 hrs then choose last 24hrs of data for baseline if
+   * it is gt 24hours then choose 2X time range for baseline calculation.
+   *
+   * @param startTimeMillis
+   * @param endTimeMillis
+   * @return
+   */
+  private long getUpdatedStartTimeForAggregations(long startTimeMillis, long endTimeMillis) {
+    if ((endTimeMillis - startTimeMillis) < DAY_IN_MILLIS) {
+      startTimeMillis = endTimeMillis - DAY_IN_MILLIS;
+    } else {
+      startTimeMillis = endTimeMillis - (2 * (endTimeMillis - startTimeMillis));
+    }
+    return startTimeMillis;
+  }
+
+  /**
+   * If user selected time range < 6 hours then 3X time range if it is > 6 hours the 2X time range.
```

Comment:
Same general comments as above - keep the data sets continuous.

Reply:
more data is always better, so unless there's performance impacts I'd go for longer. Using only 30 minutes of data to calculate a baseline seems like it defeats the purpose - any behavior change will need to be caught only if it's just beginning (e.g. if my latency spiked a half hour ago, that's already going to dominate my baseline)

Reply:
Fixed this.